File: creational-design-patterns/abstract-factory-design-pattern/Java/ui.java
import java.util.Scanner;

interface IButton {
    void press();
}

interface ITextBox {
    void setText();    
}

class MacButton implements IButton {
    @Override
    public void press() {
        System.out.println("Mac button pressed.");
    }
}

class WindowsButton implements IButton {
    @Override
    public void press() {
        System.out.println("Windows button pressed.");
    }

}
class MacTextBox implements ITextBox {
    @Override
    public void setText() {
        System.out.println("Setting text in Mac TextBox.");
    }
}

class WindowsTextBox implements ITextBox {
    @Override
    public void setText() {
        System.out.println("Setting text in Windows TextBox.");
    }
}

interface IFactory {
    IButton createButton();
    ITextBox createTextBox();
}

class WindowsFactory implements IFactory {
    @Override
    public IButton createButton() {
        return new WindowsButton();
    }

    @Override
    public ITextBox createTextBox() {
        return new WindowsTextBox();
    }
}

class MacFactory implements IFactory {
    @Override
    public IButton createButton() {
        return new MacButton();
    }

    @Override
    public ITextBox createTextBox() {
        return new MacTextBox();
    }
}

class GUIAbstractFactory {
    public static IFactory createFactory(String osType){
        if(osType.equals("windows")){
            return new WindowsFactory();
        }
        else if(osType.equals("mac")){
            return new MacFactory();
        }
        return null;
    }
}

public class ui {
    public static void main(String[] args){
        System.out.println("Enter Machine OS: ");
        Scanner scanner = new Scanner(System.in);
        String osType = scanner.nextLine();
        scanner.close();

        IFactory factory = GUIAbstractFactory.createFactory(osType);

        if(factory!=null){
            IButton button = factory.createButton();
            button.press();

            ITextBox textBox = factory.createTextBox();
            textBox.setText();
        }
        else {
            System.err.println("Invalid OS Type!!");
        }
    }
}

File: creational-design-patterns/abstract-factory-design-pattern/Typescript/ui.ts
class IButton {
    press() {}
}

class ITextBox {
    setText() {}
}

class WindowsButton extends IButton {
    press() {
        console.log("Windows button pressed.")
    }
}

class MacButton extends IButton {
    press() {
        console.log("Mac button pressed.")
    }
}

class WindowsTextBox extends ITextBox {
    setText() {
        console.log("Setting text in Windows TextBox.")
    }
}

class MacTextBox extends ITextBox {
    setText() {
        console.log("Setting text in Mac TextBox.")
    }
}

class IFactory {
    createButton(): any {}
    createTextBox(): any {}
}

class WindowsFactory {
    createButton(): IButton {
        return new WindowsButton();
    }
    createTextBox(): ITextBox {
        return new WindowsTextBox();
    }
}

class MacFactory {
    createButton(): IButton {
        return new MacButton();
    }
    createTextBox(): ITextBox {
        return new MacTextBox();
    }
}

class GUIAbstractFactory {
    static createFactory(osType: string): IFactory | null {
        if(osType == "mac"){
            return new MacFactory();
        }
        else if(osType == "windows"){
            return new WindowsFactory();
        }
        else{
            return null;
        }
    }
}

function uiMain() {
    const osType = "windows";

    const factory: IFactory | null = GUIAbstractFactory.createFactory(osType);

    if(factory!=null){
        const button: IButton = factory.createButton();
        const textBox: ITextBox = factory.createTextBox();
        button.press();
        textBox.setText();
    }
    else{
        console.log("Invalid OS Type!!");
    }
    
}

uiMain()